### ISSUE TITLE
Granian won't log or exit on ASGI lifespan startup event failures

### ISSUE THREAD

**Issue by wihlarkop** [emmett-framework/granian#147]
i got this issue when running granian with fastapi, server is running but can't open/blank in browser

![image](https://github.com/emmett-framework/granian/assets/20562624/b6e6e98b-89c4-4b2d-883a-c8134fef3f19)

my requirements.txt
```
fastapi==0.104.1
psycopg2==2.9.9
pydantic==2.4.2
pydantic-settings==2.0.3
python-dotenv==1.0.0
SQLAlchemy==2.0.23
alembic==1.12.1
granian==0.7.2
```

![image](https://github.com/emmett-framework/granian/assets/20562624/cafa3858-7100-40ce-ae06-c9a8340e4f11)

this is my main.py file
```python
class Client:
    def __init__(self):
        self.connection_url: URL = connection_url
        self.engine: Engine = create_engine(url=self.connection_url, pool_pre_ping=True)
        self.metadata = MetaData()

    def get_engine(self) -> Engine:
        return self.engine

    def disconnect(self):
        self.engine.dispose()


client = Client()

@asynccontextmanager
async def lifespan(app: FastAPI):
    db_client = client.get_engine()
    with db_client.connect() as connection:
        yield {
            "db": connection
        }
    client.disconnect()
    
app = FastAPI(
    title=settings.APP_NAME,
    version=settings.APP_VERSION,
    lifespan=lifespan,
)

if __name__ == "__main__":
    Granian(
        target="main:app",
        interface=Interfaces.ASGI,
        reload=True
    ).serve()
```

and i run
```python main.py```

**Comment by ameyer117** [comment]
I believe this is because Granian doesn't have lifespan support yet.

[You can use the old startup events as a workaround](https://fastapi.tiangolo.com/advanced/events/#alternative-events-deprecated)

**Comment by gi0baro** [comment]
@wihlarkop @ameyer117 yep I suppose this needs #94, but I didn't check if the `lifespan` parameter of starlette (and thus FastAPI) requires that. Probably someone like @Kludex can confirm this.

**Comment by wihlarkop** [comment]
Hi @Kludex 

there is no log or error when i start the server, when i access 127.0.0.1:8000 it just blank page like my ss and in cli there is no error or warning too, even when i add log_enabled=True in granian still no error log
![image](https://github.com/emmett-framework/granian/assets/20562624/7db50163-6d07-4346-9c5f-8766ff163e3f)

**Comment by wihlarkop** [comment]
@ameyer117 it can work with your solution , but on_event workaround will deprecated so i want to migrate to lifespan event handlers

**Comment by wihlarkop** [comment]
@Kludex 
in my case granian just give log like this

```
[INFO] Starting granian
[INFO] Listening at: 127.0.0.1:8000
[INFO] Spawning worker-1 with pid: 10501
```

if you want try my code you can access this https://github.com/wihlarkop/fastapi-granian

**Comment by gi0baro** [comment]
Hey all, I debugged this locally.

So there are two themes here:
- Granian doesn't support lifespan state yet, it will be included with #94 
- Starlette actually behaves correctly as it actually rises a RuntimeError here https://github.com/encode/starlette/blob/93494a46ee0f2e8b1eeafd12e466ebac15037b41/starlette/routing.py#L708 as there's no lifespan state in the protocol

The actual issue here is that Granian is actually catching the exception and exiting the worker process without logging or re-raising anything, and since there's no fine-grained process management today in granian, the main process is kept alive even if the actual worker prematurely exited (this is a theme also in #148 ).

So in conclusion:
- @wihlarkop the solution for your issue is actually to use old startup events until granian implements lifespan state (targeted for next revision)
- @Kludex everything is correct on the starlette side of things, I'm sorry for the bother
- I gonna keep this open but change the title a bit to keep track of the logging/exit issue, which should be fixed ASAP
